src/common/webrequest_curl.cpp

```diff
@@ -419,160 +927,357 @@ wxWebSessionCURL::CreateRequest(wxWebSession& session,
+            }
+        }
+    }
+}
+
+void wxWebSessionCURL::FailRequest(CURL* curl,const wxString& msg)
 {
-    // Add the request to a list of threads that will be removed from the curl
-    // multi handle in the worker thread
-    wxCriticalSectionLocker lock(m_cancelled.cs);
-    request->IncRef();
-    m_cancelled.requests.push_back(wxObjectDataPtr<wxWebRequestCURL>(request));
+    TransferSet::iterator it = m_activeTransfers.find(curl);
+
+    if ( it != m_activeTransfers.end() )
+    {
+        wxWebRequestCURL* request = it->second;
+        m_activeTransfers.erase(it);
+        curl_multi_remove_handle(m_handle, curl);
+        StopTransfer(curl);
+
+        request->SetState(wxWebRequest::State_Failed, msg);
+    }
 }
 
-wxVersionInfo  wxWebSessionCURL::GetLibraryVersionInfo()
+void wxWebSessionCURL::StopTransfer(CURL* curl)
 {
-    const curl_version_info_data* vi = curl_version_info(CURLVERSION_NOW);
-    wxString desc = wxString::Format("libcurl/%s", vi->version);
-    if (vi->ssl_version[0])
-        desc += " " + wxString(vi->ssl_version);
-    return wxVersionInfo("libcurl",
-        vi->version_num >> 16 & 0xff,
-        vi->version_num >> 8 & 0xff,
-        vi->version_num & 0xff,
-        desc);
+    curl_socket_t activeSocket;
+    bool closeActiveSocket = true;
+    bool useLastSocket = false;
+
+#if CURL_AT_LEAST_VERSION(7, 45, 0)
+    if ( CurlRuntimeAtLeastVersion(7, 45, 0) )
+    {
+        CURLcode code = curl_easy_getinfo(curl, CURLINFO_ACTIVESOCKET,
+                                          &activeSocket);
+
+        if ( code != CURLE_OK || activeSocket == CURL_SOCKET_BAD )
+        {
+            closeActiveSocket = false;
+        }
+    }
+    else
+    {
+        useLastSocket = true;
+    }
+#else
+    useLastSocket = true;
+#endif //CURL_AT_LEAST_VERSION(7, 45, 0)
+
+    // CURLINFO_ACTIVESOCKET is not available either at compile time or run
+    // time. So we must use the older CURLINFO_LASTSOCKET instead.
+    if ( useLastSocket )
+    {
+        #ifdef __WIN64__
+            // CURLINFO_LASTSOCKET won't work on 64 bit windows because it
+            // uses a long to retrive the socket. However sockets will be 64
+            // bit values. In this case there is nothing we can do.
+            closeActiveSocket = false;
+        #endif //__WIN64__
+
+        if ( closeActiveSocket )
+        {
+            long longSocket;
+            CURLcode code = curl_easy_getinfo(curl, CURLINFO_LASTSOCKET,
+                                              &longSocket);
+
+            if ( code == CURLE_OK && longSocket!= -1 )
+            {
+                activeSocket = static_cast<curl_socket_t>(longSocket);
+            }
+            else
+            {
+                closeActiveSocket = false;
+            }
+        }
+    }
+
+    if ( closeActiveSocket )
+    {
+        #ifdef __WINDOWS__
+            closesocket(activeSocket);
+        #else
+            close(activeSocket);
+        #endif
```

Comment:
```suggestion
        wxCloseSocket(activeSocket);
```
(this requires including `wx/private/socket.h`)